File: baseLib/src/main/java/me/ycdev/android/lib/common/dbmgr/SQLiteDbMgr.java
package me.ycdev.android.lib.common.dbmgr;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import androidx.annotation.NonNull;

import java.util.HashMap;

import me.ycdev.android.lib.common.utils.LibConfigs;
import me.ycdev.android.lib.common.utils.LibLogger;

@SuppressWarnings("unused")
public class SQLiteDbMgr {
    private static final String TAG = "SQLiteDbMgr";
    private static final boolean DEBUG = LibConfigs.DEBUG_LOG;

    private static class DbInfo {
        SQLiteDatabase db;
        int referenceCount;
    }

    private Context mAppContext;
    private HashMap<Class<? extends SQLiteDbCreator>, DbInfo> mOpenHelpers = new HashMap<>();

    @SuppressLint("StaticFieldLeak")
    private static volatile SQLiteDbMgr sInstance;

    private SQLiteDbMgr(Context cxt) {
        mAppContext = cxt.getApplicationContext();
    }

    private static SQLiteDbMgr getInstance(Context cxt) {
        if (sInstance == null) {
            synchronized (SQLiteDbMgr.class) {
                if (sInstance == null) {
                    sInstance = new SQLiteDbMgr(cxt);
                }
            }
        }
        return sInstance;
    }

    private SQLiteDatabase acquireDatabase(Class<? extends SQLiteDbCreator> dbInfoClass) {
        if (DEBUG) LibLogger.d(TAG, "acquire DB: " + dbInfoClass.getName());
        SQLiteDatabase db;
        synchronized (SQLiteDbMgr.class) {
            DbInfo info = mOpenHelpers.get(dbInfoClass);
            if (info == null) {
                try {
                    if (DEBUG) LibLogger.d(TAG, "create DB: " + dbInfoClass.getName());
                    SQLiteDbCreator helper = dbInfoClass.newInstance();
                    info = new DbInfo();
                    info.db = helper.createDb(mAppContext);
                    info.referenceCount = 0;
                    mOpenHelpers.put(dbInfoClass, info);
                } catch (Exception e) {
                    throw new RuntimeException("failed to create SQLiteOpenHelper instance", e);
                }
            }
            info.referenceCount++;
            db = info.db;
        }
        return db;
    }

    private void releaseDatabase(Class<? extends SQLiteDbCreator> dbInfoClass) {
        if (DEBUG) LibLogger.d(TAG, "release DB: " + dbInfoClass.getName());
        synchronized (SQLiteDbMgr.class) {
            DbInfo info = mOpenHelpers.get(dbInfoClass);
            if (info != null) {
                info.referenceCount--;
                if (info.referenceCount == 0) {
                    if (DEBUG) LibLogger.d(TAG, "close DB: " + dbInfoClass.getName());
                    info.db.close();
                    info.db = null;
                    mOpenHelpers.remove(dbInfoClass);
                }
            }
        }
    }

    public static SQLiteDatabase acquireDatabase(@NonNull Context cxt,
            @NonNull Class<? extends SQLiteDbCreator> dbInfoClass) {
        return getInstance(cxt).acquireDatabase(dbInfoClass);
    }

    public static void releaseDatabase(@NonNull Context cxt,
            @NonNull Class<? extends SQLiteDbCreator> dbInfoClass) {
        getInstance(cxt).releaseDatabase(dbInfoClass);
    }
}

File: baseLib/src/main/java/me/ycdev/android/lib/common/dbmgr/SQLiteDbMgr.kt
package me.ycdev.android.lib.common.dbmgr

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import me.ycdev.android.lib.common.pattern.SingletonHolderP1
import timber.log.Timber
import java.util.HashMap

@Suppress("unused")
class SQLiteDbMgr private constructor(cxt: Context) {

    private val mAppContext: Context = cxt.applicationContext
    private val mOpenHelpers = HashMap<Class<out SQLiteDbCreator>, DbInfo>()

    private class DbInfo {
        internal var db: SQLiteDatabase? = null
        internal var referenceCount: Int = 0
    }

    private fun acquireDatabase(dbInfoClass: Class<out SQLiteDbCreator>): SQLiteDatabase? {
        Timber.tag(TAG).d("acquire DB: %s", dbInfoClass.name)
        val db: SQLiteDatabase?
        synchronized(SQLiteDbMgr::class.java) {
            var info = mOpenHelpers[dbInfoClass]
            if (info == null) {
                try {
                    Timber.tag(TAG).d("create DB: %s", dbInfoClass.name)
                    val helper = dbInfoClass.newInstance()
                    info = DbInfo()
                    info.db = helper.createDb(mAppContext)
                    info.referenceCount = 0
                    mOpenHelpers[dbInfoClass] = info
                } catch (e: Exception) {
                    throw RuntimeException("failed to create SQLiteOpenHelper instance", e)
                }
            }
            info.referenceCount++
            db = info.db
        }
        return db
    }

    private fun releaseDatabase(dbInfoClass: Class<out SQLiteDbCreator>) {
        Timber.tag(TAG).d("release DB: %s", dbInfoClass.name)
        synchronized(SQLiteDbMgr::class.java) {
            val info = mOpenHelpers[dbInfoClass]
            if (info != null) {
                info.referenceCount--
                if (info.referenceCount == 0) {
                    Timber.tag(TAG).d("close DB: %s", dbInfoClass.name)
                    info.db!!.close()
                    info.db = null
                    mOpenHelpers.remove(dbInfoClass)
                }
            }
        }
    }

    companion object : SingletonHolderP1<SQLiteDbMgr, Context>(::SQLiteDbMgr) {
        private const val TAG = "SQLiteDbMgr"

        fun acquireDatabase(
            cxt: Context,
            dbInfoClass: Class<out SQLiteDbCreator>
        ): SQLiteDatabase? {
            return getInstance(cxt).acquireDatabase(dbInfoClass)
        }

        fun releaseDatabase(
            cxt: Context,
            dbInfoClass: Class<out SQLiteDbCreator>
        ) {
            getInstance(cxt).releaseDatabase(dbInfoClass)
        }
    }
}
